### ISSUE TITLE
C++ API to close/disconnect a grpc::Channel, canceling all calls

### ISSUE THREAD

**Issue by mehrdada** [grpc/grpc#21926]
`grpc_channel_destroy` C API cancels all active outgoing calls on the channel. This API is invoked in C++ `grpc::Channel` destructor. However, when a new call is created, a *strong* reference to channel `shared_ptr` gets added to the Call object, making it a circular reference to the `Channel` object, leaving no way for channel destruction.  This API would be necessary when you want to cleanly shut down some process and unblock threads that are waiting for calls to proceed, but cannot. Barring that, the user would have to manually keep track of all calls associated with the channel which is unnecessary and wasteful. This should either happen automatically by losing the final reference to the `shared_ptr<Channel>` or via an explicit `Close()` API.

**Comment by markdroth** [comment]
Actually, `grpc_channel_destroy()` does _not_ cancel all active outgoing calls on the channel.  Despite its name, its behavior is really more like `grpc_channel_unref()` -- all it does it drop a ref, so that the channel will be destroyed when all the calls complete.  See #14292 for related discussion.

I'll leave this open as a separate issue, because even if #14292 gets fixed at some point, the extra ref being held to the C++ object would also need to be addressed.

Note that this will be fairly complicated to fix, and we have other priorities at the moment, so I wouldn't expect this to be addressed anytime soon.

**Comment by MaxBrinegar** [comment]
Any update on this?

**Comment by markdroth** [comment]
No, there's been no work on this.  It's not likely to get any attention any time soon, since we have so many other higher priority projects.

**Comment by rok** [comment]
Any updates on priorities to share?

**Comment by rok** [comment]
@markdroth would you be open to a PR for this and #14292 if needed?

**Comment by markdroth** [comment]
Note that applications can do this themselves by maintaining a list of all calls they start on a channel and, when they're ready to shut down the channel, cancelling all of those calls.  So there is a work-around here.

**Comment by ctiller** [comment]
I think the trick here would be to move *all* of the cost to the cancellation point (perhaps introducing a `grpc_channel_cancel_all_calls()` method analogous to `grpc_server_cancel_all_calls()`).

Which is to say, an absolute requirement here would be no cost added to the usual call path.

**Comment by amandeepgautam** [comment]
Any updates on this?
